File: Week 1 - C/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

bool check_log(long cc);
int check_length(long cc);
int check_front(long cc);
void identify(long cc, int length, int front_digits);

int main(void)
{
    // request input
    long cc = get_long("Your CC-Number: ");

    // call functions, if Luhn's Algorithm equals true
    if (check_log(cc))
    {
        int length = check_length(cc);
        int front = check_front(cc);
        identify(cc, length, front);
    }
    else
    {
        printf("INVALID\n");
/*         printf("Invalid card number.\n"); */
    }
}

// Check Luhn's Algorithm
bool check_log(long cc)
{
    int sum = 0;
    bool straight_pos = false;
    while (cc > 0)
    {
        int last_digit = cc % 10;

        if (straight_pos == true)
        {
            if (last_digit >= 5)
            {
                int digits = last_digit * 2;
                while (digits > 0)
                {
                    int last = digits % 10;
                    sum += last;
                    digits = digits / 10;
                }
            }
            else
            {
                sum += last_digit * 2;
            }
        }
        else
        {
            sum += last_digit;
        }
        straight_pos = !straight_pos;
        cc = cc / 10;
    }

    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

// Check Length of inserted Card Number
int check_length(long cc)
{
    int length = 0;
    while (cc > 0)
    {
        length++;
        cc = cc / 10;
    }
    return length;
}

// Check the first 2 digits
int check_front(long cc)
{
    int front_digits;
    while (cc > 0)
    {
        if (cc < 100 && cc > 10)
        {
            front_digits = cc;
        }
        cc = cc / 10;
    }
    return front_digits;
}

// Combine everything & give feedback
void identify(long cc, int length, int front_digits)
{
    if (length == 15 && (front_digits == 34 || front_digits == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && front_digits >= 51 && front_digits <= 55)
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && front_digits >= 40 && front_digits <= 49)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week 6 - Python/Credit/credit.py
from sys import exit

# CS50 Problem set 1's "credit"


# get_int function (-> no import of cs50 needed)
def get_num(text):
    while True:
        try:
            return int(input(text))
        except:
            print("Please type in a number!")


# check length and luhn's algorithm
def check(num, length):
    # check cardnumber length
    if length < 13 or length > 16:
        print("INVALID")
        exit(f"Invalid card length. Was {length} (Min 13, max 16 digits)")
    # else calc luhn's algorithm
    else:
        sum = 0
        for i in range(length):
            if i % 2 == 0:
                sum += num % 10
            else:
                odd = (num % 10) * 2
                sum += odd // 10 + odd % 10
            num //= 10
        if sum % 10 == 0:
            return True
        else:
            print("INVALID")
            exit("Card number did not match algorithm.")


# identify company
def identify(card, length):
    # str for iteration and back to int for calculation
    dig1 = int(str(card)[0])
    dig2 = int(str(card)[1])

    if length == 15 and dig1 == 3 and (dig2 == 4 or dig2 == 7):
        return "AMEX"
    elif length == 16 and dig1 == 5 and dig2 >= 1 and dig2 <= 5:
        return "MASTERCARD"
    elif (length == 13 or length == 16) and dig1 == 4 and dig2 >= 0 and dig2 <= 9:
        return "VISA"
    else:
        print("INVALID")
        exit("No company matches found.")


# define main function
def main():
    card = get_num("Your Card Number: ")
    length = len(str(card))
    # everything is okay? -> Print company name
    if check(card, length):
        print(identify(card, length))
    # exit() = code 0
    exit()


# call main function if it's not a module
if __name__ == "__main__":
    main()
